registry/storage/driver/middleware/cloudfront/middleware.go

```diff
@@ -114,8 +130,66 @@ type S3BucketKeyer interface {
 	S3BucketKey(path string) string
 }
 
-// Resolve returns an http.Handler which can serve the contents of the given
-// Layer, or an error if not supported by the storagedriver.
+// parseIPFromRequest attempts to extract the ip address of the
+// client that made the request
+func parseIPFromRequest(ctx context.Context) (net.IP, error) {
+	request, err := dcontext.GetRequest(ctx)
+	if err != nil {
+		return nil, err
+	}
+	ipStr := dcontext.RemoteIP(request)
+	ip := net.ParseIP(ipStr)
+	if ip == nil {
+		return nil, fmt.Errorf("invalid ip address from requester: %s", ipStr)
+	}
+
+	return ip, nil
+}
+
+// Client facing behavior changed slightly with the rollout of the s3filter.
+// Some clients, such as curl, automatically follow redirects from the v2/image/blobs/someshasum
+// endpoint. This is problematic if the 'Authorization' header, meant for registry, is forwarded
+// to the storage backend. S3 also has an 'Authorization' header, which can be used in place
+// of a signed url. If the client forwards the registry 'Authorization' header to s3,
+// s3 gets confused and fails to return the data.
+//
+// This function recognizes 'proper' clients based on the user agent. To avoid breaking clients
+// such as curl, we will only return s3 urls for proper clients.
+//
+// This function currently allows docker clients >= 1.6.0
+func clientDropsAuthHeader(ctx context.Context) bool {
+	request, err := dcontext.GetRequest(ctx)
+	if err != nil {
+		dcontext.GetLogger(ctx).Warnf("s3filter: No request object in context")
+		return false
+	}
+	ua := request.UserAgent()
+	if !strings.HasPrefix(ua, "docker/") {
+		return false
+	}
+
+	// A docker user agent may have more than the semver.
+	// ex: 'docker/17.03.1-ee-2 go/go1.7'
+	// Where only '17.03.1-ee-2' is a semver.
+	versionParts := strings.Split(strings.TrimPrefix(ua, "docker/"), " ")
+	if len(versionParts) == 0 {
+		// malformed docker version
+		return false
+	}
+	v, err := version.NewVersion(versionParts[0])
+	if err != nil {
+		dcontext.GetLogger(ctx).Errorf("Failed to parse semver from useragent='%s' : %s", ua, err.Error())
```

Comment:
Don't capitalize log messages.

Reply:
Also, please use `WithError` and `WithField` to decorate the error and user agent.